### PR TITLE
NMS-14119: Implement support for SSH host key authentication

### PR DESCRIPTION
**Pull request by fooker**
### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-14119

### CODE REVIEW

File: features/device-config/ssh-scripting/src/main/java/org/opennms/features/deviceconfig/sshscripting/impl/SshScriptingServiceImpl.java [hunk 1]
```diff
@@ -190,7 +198,22 @@ private SshInteractionImpl(
                 }
 
                 try {
-                    session.addPasswordIdentity(password);
+                    if (password != null) {
+                        session.addPasswordIdentity(password);
+                    }
+
+                    if (!Strings.isNullOrEmpty(authKey)) {
```
Comment (cgorantla) [inline features/device-config/ssh-scripting/src/main/java/org/opennms/features/deviceconfig/sshscripting/impl/SshScriptingServiceImpl.java:205]: If it is possible to add a test for this ?

Comment (fooker) [reply features/device-config/ssh-scripting/src/main/java/org/opennms/features/deviceconfig/sshscripting/impl/SshScriptingServiceImpl.java:201]: Tests added

File: features/device-config/ssh-scripting/pom.xml [hunk 2]
```diff
@@ -51,6 +51,10 @@
             <groupId>org.opennms</groupId>
             <artifactId>opennms-util</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.bouncycastle</groupId>
```
Comment (cgorantla) [inline features/device-config/ssh-scripting/pom.xml:55]: So we can remove this dependency now ?

File: docs/modules/operation/pages/device-config-backup/dcb-requisition.adoc [hunk 3]
```diff
@@ -16,10 +16,10 @@ If you want to enable device config backup on existing nodes in a requisition, s
 . Click on the edit icon and click *Add Node*.
 . Set up your requisition as described in xref:operation:provisioning/getting-started.adoc#requisition-create[Create a requisition].
 . Click *Add Meta-Data*, and define the following key-value pairs:
-.. `username`: username for the device
-.. `password`: password for the device
-.. `script-file`: name of the device configuration script for this type of device
-.. `schedule`: a cron expression to specify the time to fetch the device backup configuration (default is `0 0 0 * * ?`, every 24 hrs at 12:00 a.m.)
+.. `dcb-username`: username for the device
```
Comment (cgorantla) [inline docs/modules/operation/pages/device-config-backup/dcb-requisition.adoc:19]: Should we also update default poller configuration 
They are currently `dcb:username`.

**Review by cgorantla — CHANGES_REQUESTED** [review]

File: features/device-config/ssh-scripting/pom.xml (hunk 2 above)
Comment (fooker) [reply features/device-config/ssh-scripting/pom.xml:55]: removed

**Review by cgorantla — APPROVED** [review]
LGTM!

File: docs/modules/operation/pages/device-config-backup/dcb-requisition.adoc [hunk 4]
```diff
@@ -16,10 +16,10 @@ If you want to enable device config backup on existing nodes in a requisition, s
...
 . Set up your requisition as described in xref:operation:provisioning/getting-started.adoc#requisition-create[Create a requisition].
 . Click *Add Meta-Data*, and define the following key-value pairs:
-.. `username`: username for the device
-.. `password`: password for the device
-.. `script-file`: name of the device configuration script for this type of device
-.. `schedule`: a cron expression to specify the time to fetch the device backup configuration (default is `0 0 0 * * ?`, every 24 hrs at 12:00 a.m.)
+.. `dcb:username`: username for the device
+.. `dcb:password`: password for the device
+.. `dcb:script-file`: name of the device configuration script for this type of device
+.. `dcb:schedule`: a cron expression to specify the time to fetch the device backup configuration (default is `0 0 0 * * ?`, every 24 hrs at 12:00 a.m.)
```
Comment (Bonrob2) [inline docs/modules/operation/pages/device-config-backup/dcb-requisition.adoc:22]: Just to verify that this is actually dcb:username? Chandra had said to remove "dcb".

**Review by Bonrob2 — DISMISSED** [review]
Approving docs so as not to gate merge. But please note my comment and update if required.

File: docs/modules/operation/pages/device-config-backup/dcb-requisition.adoc (hunk 4 above)
Comment (cgorantla) [reply docs/modules/operation/pages/device-config-backup/dcb-requisition.adoc:22]: @Bonrob2  yes, this is fine. I wasn't sure if we keep the same naming when we implement SCV. 
Also this naming is not something that is hard-coded.  This naming is referred in poller-configuration.xml, So user can update these.

**Review by cgorantla — APPROVED** [review]

File: docs/modules/operation/pages/device-config-backup/dcb-requisition.adoc (hunk 4 above)
Comment (Bonrob2) [reply docs/modules/operation/pages/device-config-backup/dcb-requisition.adoc:22]: Thanks for confirming!

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfigMonitor.adoc [hunk 5]
```diff
@@ -45,6 +45,11 @@ The default cron expression runs every 24 hrs at 12:00 a.m.
   One can use `ssh-keygen -l -f <(ssh-keyscan $TARGET)` to find the fingerprints of a remote system.
 | _unchecked_
 
+| auth-key
+| SSH Private key used to authenticate.
+  This must be specified as PEM file.
```
Comment (christianpape) [inline docs/modules/reference/pages/service-assurance/monitors/DeviceConfigMonitor.adoc:50]: Please user format instead of file here.

**Review by christianpape — APPROVED** [review]
LGTM.

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfigMonitor.adoc [hunk 6]
```diff
@@ -45,6 +45,11 @@ The default cron expression runs every 24 hrs at 12:00 a.m.
   One can use `ssh-keygen -l -f <(ssh-keyscan $TARGET)` to find the fingerprints of a remote system.
 | _unchecked_
```
Comment (mmahacek) [inline docs/modules/reference/pages/service-assurance/monitors/DeviceConfigMonitor.adoc:46]:
```suggestion
| n/a
```

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfigMonitor.adoc [hunk 7]
```diff
@@ -45,6 +45,11 @@ The default cron expression runs every 24 hrs at 12:00 a.m.
...
 
+| auth-key
+| SSH Private key used to authenticate.
+  This must be specified in PEM format.
+| _unused_
+
```
Comment (mmahacek) [inline docs/modules/reference/pages/service-assurance/monitors/DeviceConfigMonitor.adoc:52]:
```suggestion
| n/a
```

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfigMonitor.adoc (hunk 6 above)
Comment (fooker) [reply docs/modules/reference/pages/service-assurance/monitors/DeviceConfigMonitor.adoc:46]: I don't think `n/a` is a good match. The default value really means, that the host key is not checked.

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfigMonitor.adoc (hunk 7 above)
Comment (fooker) [reply docs/modules/reference/pages/service-assurance/monitors/DeviceConfigMonitor.adoc:52]: Same here. If no key is specified, key auth is not tried.

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfigMonitor.adoc (hunk 6 above)
Comment (Bonrob2) [reply docs/modules/reference/pages/service-assurance/monitors/DeviceConfigMonitor.adoc:46]: Is there a default value, or is this more of a default action? Would it be clearer to say "By default, the monitor does not check the host key."

**Comment by fooker** [comment]
@Bonrob2 I've added these changes to the description of the fields. Is this sufficient?

**Comment by fooker** [comment]
> LGTM !
> 
> Should we update document on what `shell` parameter supposed to mean ?

Docs added for `shell`

**Review by Bonrob2 — COMMENTED** [review]
Minor change. Otherwise, looks good.

**Comment by Bonrob2** [comment]
> @Bonrob2 I've added these changes to the description of the fields. Is this sufficient?

yes, thanks!

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfigMonitor.adoc [hunk 8]
```diff
@@ -77,8 +77,19 @@ Example: `P20Y2M25D` means 20 years, 2 months, and 25 days.
...
 | _unchecked_
 
+| auth-key
+| SSH Private key used to authenticate.
+  This must be specified in PEM format.
+  If unspecified, key authentication is not used.
+| _unused_
+
+| shell
+| If specified, this command is executed on the remote host instead of a login shell.
```
[the remark this answers is not in the document]
Comment (Bonrob2) [reply docs/modules/reference/pages/service-assurance/monitors/DeviceConfigMonitor.adoc:90]: Fair enough. I think it's better not to use the passive voice. Is there something that executes the command. That is not clear with the way it is written. But I'm fine to keep the original if it really doesn't matter.

**Review by cgorantla — APPROVED** [review]

**Review by christianpape — APPROVED** [review]
LGTM.

**Pull request merged**